pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -150,17 +169,67 @@ func (state PipelineRunState) IsDone() (isDone bool) {
 // GetNextTasks will return the next ResolvedPipelineRunTasks to execute, which are the ones in the
 // list of candidateTasks which aren't yet indicated in state to be running.
 func (state PipelineRunState) GetNextTasks(candidateTasks map[string]struct{}) []*ResolvedPipelineRunTask {
+	// initialize a list of tasks to return tasks next in execution queue
 	tasks := []*ResolvedPipelineRunTask{}
+	// iterate over PipelineRunState to get TaskRun and PipelineTask for each candidate tasks
 	for _, t := range state {
-		if _, ok := candidateTasks[t.PipelineTask.Name]; ok && t.TaskRun == nil {
-			tasks = append(tasks, t)
-		}
-		if _, ok := candidateTasks[t.PipelineTask.Name]; ok && t.TaskRun != nil {
-			status := t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
-			if status != nil && status.IsFalse() {
-				if !(t.TaskRun.IsCancelled() || status.Reason == v1alpha1.TaskRunSpecStatusCancelled || status.Reason == ReasonConditionCheckFailed) {
-					if len(t.TaskRun.Status.RetriesStatus) < t.PipelineTask.Retries {
-						tasks = append(tasks, t)
+		// for each candidate task, check if its TaskRun is initialized or not
```

Comment:
The comments are super helpful! Thanks! 
The nesting here is quite deep. Can we extract some bits onto helper functions? e.g  the `if t.TaskRun == nil` branch can be its own function probably..

Reply:
Done :)